cache/network/api.go

```diff
@@ -140,24 +155,156 @@ func (c apiClient) uploadArchive(archivePath, uploadMethod, uploadURL string, he
 	if err != nil {
 		c.logger.Warnf("error while dumping response: %s", err)
 	}
-	c.logger.Debugf("Response dump: %s", string(dump))
+	c.logger.Debugf("Chunk response dump: %s", string(dump))
 
 	if resp.StatusCode != http.StatusOK {
-		return unwrapError(resp)
+		return "", unwrapError(resp)
+	}
+
+	etag := resp.Header.Get("ETag")
+	fmt.Printf("Etag %s for %s of size %v \n\r", etag, uploadURL.URL, size)
+
+	return etag, nil
+}
+
+func (c apiClient) uploadArchive(archivePath string, chunkSize, chunkCount, lastChunkSize int, uploadURLs []uploadURL) ([]string, error) {
+
+	fmt.Printf("Uploading archive: %s \n\r", archivePath)
+	fmt.Printf("Chunk size, count, last chunk size: %d, %d, %d \n\r", chunkSize, chunkCount, lastChunkSize)
+	fmt.Printf("Upload urls: %+v \n\r", uploadURLs)
+
+	file, err := os.Open(archivePath)
+	if err != nil {
+		return nil, fmt.Errorf("open file: %s", err)
+	}
+	defer file.Close()
+
+	if chunkCount == 1 {
+		fileInfo, err := os.Stat(archivePath)
+		if err != nil {
+			return nil, fmt.Errorf("stat file: %s", err)
+		}
+		c.logger.Debugf("Uploading single chunk (non multipart upload)", uploadURLs[0].URL)
+
+		_, err = c.uploadArchiveChunk(uploadURLs[0], file, fileInfo.Size())
+		if err != nil {
+			return nil, fmt.Errorf("upload single chunk (non multipart): %s", err)
+		}
+		return nil, nil
+	}
+
+	type Job struct {
+		ChunkNumber int
+		ChunkStart  int64
+		ChunkSize   int64
+		UploadURL   uploadURL
+	}
+
+	jobs := make(chan Job, chunkCount)
+	results := make(chan struct {
+		ChunkNumber int
+		ChunkSize   int64
+		ETag        string
+	}, chunkCount)
+	errors := make(chan error, chunkCount)
+	wg := sync.WaitGroup{}
+
+	worker := func() {
+		for job := range jobs {
+
+			chunkData, err := io.ReadAll(io.NewSectionReader(file, job.ChunkStart, job.ChunkSize))
+			if err != nil {
+				errors <- fmt.Errorf("read chunk %d: %s", job.ChunkNumber, err)
+				wg.Done()
```

Comment:
Use `defer()` instead of adding this to all lines

Reply:
Unless if you change the structure to a select-based cancelable one. In that case only call done if a piece of chunk has been processed

Reply:
Even better yet - use the WaitGroup only for synchronizing the goroutines, since we're waiting on the goroutines, they can handle their own shutdown when there are no more items in the channel. So like
```
var wg sync.WaitGroup

ctx, cancelFunc := context.WithCancel(context.Background()) // Or if we already have a parent, use that
ch := make(chan int, chunkCount)
errCh := make(chan error)
for i := 0; i < chunkCount; i++ {
	ch <- job
}
close(ch)
for i := 0; i < 10; i++ {
	wg.Add(1)
	go func() {
		defer wg.Done()
		for {
			select {
			case job, ok := <-ch:
				if !ok {
					return
				}
				if err := uploadChunk(job); err != nil {
					cancelFunc()
					return
				}
			case <-ctx.Done():
			default:
				return
			}
		}
	}()
}
wg.Wait() // Wait on workers
```